src/runtime-rs/crates/hypervisor/src/device/topology.rs

```diff
@@ -648,6 +655,31 @@ impl PCIeTopology {
 
         Ok(())
     }
+
+    /// Finds an availabled node in the PCIe topology.
+    /// Returns the first available node found, either a TopologyPortDevice or a SwitchDownPort.
+    pub fn find_available_node(&mut self) -> Option<AvailableNode> {
+        // search in pcie_port_devices
+        for port_device in self.pcie_port_devices.values_mut() {
+            if !port_device.allocated {
+                port_device.allocated = true;
+                return Some(AvailableNode::TopologyPortDevice(port_device.clone()));
+            }
+
+            // search in connected switch's downstream ports
+            if let Some(switch) = &mut port_device.connected_switch {
+                for switch_port in switch.switch_ports.values_mut() {
+                    if !switch_port.allocated {
+                        switch_port.allocated = true;
+                        return Some(AvailableNode::SwitchDownPort(switch_port.clone()));
+                    }
+                }
+            }
```

Comment:
Just out of curiosity and seeing as we don't search recursively here, is it the case that a switch cannot be attached to a down port of another switch?

Reply:
Thx Pavel, good cache! In my aspect, we should ensure support enough functions but  also take the lightweight factor into account. If complex pcie topology is indeed needed and can bring much more benefits, I think we should support such ability allowing one switch attached to another's swith downstream port. To be honest, I am not sure if these cases should be supported in our lightweight VM-based containers.

Reply:
Sure, if we're making a conscious decision here not to support more complex topologies since we have a reason to believe that we don't need them, that's fair enough for me.